### DISCUSSION TITLE
Build and upload multi-platform Docker images

### DISCUSSION THREAD

**Discussion by jehrhardt** [CardanoSolutions/ogmios#260] [Archives]
#### Describe your idea, in simple words.
Publish multi-platform Docker images to Docker Hub. Support at least:

- linux/amd64
- linux/arm64

#### Why is it a good idea?
AWS provides Graviton as a cheaper option. Apple provides its laptops with M1 / M2 chips. Multi-platform images can provide support for multiple CPU architectures in a single image.

#### Are you willing to work on it yourself?
🤔 probably yes.

**Comment by jehrhardt** [comment]
The pull requests added arm64 support to the ogmios image, but not to the one bundled with Cardano node. Is this enough? Or should we investigate to build the other for arm64 as well?

**Comment by KtorZ** [comment]
I wonder if it's possible. I haven't tried to build it and I am far away from my laptop for the next week so I can't try. For the sake of consistency and completeness, I'd rather have it if possible. The cardano-node-ogmios' image however is based on the cardano-node upstream image, which may or may not play well with this cross-platform docker builds.

**Comment by jehrhardt** [comment]
Ah, damn. Cardano node upstream image is amdzr only. Maybe we should close this one and follow up, once the upstream image works.

**Comment by KtorZ** [comment]
The build is failing for arm for the base image as well 😔. If you want to have a look, otherwise I'll next week.

**Comment by wolf31o2** [comment]
I publish cardano-node images built from upstream source, but on both AMD64 and ARM64, with the same interface as the upstream IOG images. I was just about to create my own wrapper for this project so I could get working ARM64 images. Would you accept a pull request that switched the images? It would be a drop-in replacement in this usage.

Images:
https://hub.docker.com/r/cloudstruct/cardano-node
Sources:
https://github.com/cloudstruct/docker-cardano-node

**Reply by KtorZ** [reply]
That's a lovely idea. I'll have to be _slightly careful_ however and probably fork the cloudstruct repository. For something as sensitive as package distribution, I need to keep that under the same organization as Ogmios.

**Comment by pedromtcosta** [comment]
What is the current status of this topic? Is it possible to run Ogmios on the Apple M1 / M2 chips at the moment?

**Reply by KtorZ** [reply]
It is possible, but you have to compile from arm64 (I am running and developing on M1 without issue). 

I've tried cross compiling from amd but without success; upstream dependencies (in particular Plutus) do not play nicely with it. There's no arm runner available on GitHub so I am considering setting up a self-hosted arm runner but it hasn't happened yet and I don't know when it will.

**Reply by pedromtcosta** [reply]
Could you give me some guidance on how I can compile it using arm64? I tried changing all references from amd64 to arm64 on the Dockerfile and also changing this part on `default.nix` to arm64:
```
platform = {
    arm64 = (mkProject pkgs.pkgsCross.aarch64-multiplatform-musl);
    # amd64 = (mkProject pkgs.pkgsCross.musl64);
  };
```
but then when I run `docker build` I get this error:
```
#15 2512.1 make[1]: *** [compiler/ghc.mk:165: compiler/stage2/build/HsInstances.p_o] Killed
#15 2512.1 make[1]: *** Waiting for unfinished jobs....
#15 2512.1 <<ghc: 769501104 bytes, 106 GCs, 9002056/28059256 avg/max bytes residency (7 samples), 59M in use, 0.001 INIT (0.007 elapsed), 0.216 MUT (1.921 elapsed), 0.114 GC (1.068 elapsed) :ghc>>
#15 2512.1 <<ghc: 15889829328 bytes, 790 GCs, 65797887/135834528 avg/max bytes residency (21 samples), 352M in use, 0.000 INIT (0.001 elapsed), 3.453 MUT (15.164 elapsed), 2.313 GC (2.514 elapsed) :ghc>>
#15 2512.1 <<ghc: 28405562616 bytes, 641 GCs, 167537036/828120768 avg/max bytes residency (17 samples), 1980M in use, 0.000 INIT (0.000 elapsed), 6.007 MUT (20.860 elapsed), 4.937 GC (5.751 elapsed) :ghc>>
#15 2512.1 make: *** [Makefile:128: all] Error 2
#15 2512.1 error: builder for '/nix/store/d76p47xr809y3874a9szrf3sfs3rks00-ghc-8.8.4.drv' failed with exit code 2
```
